x/feedist/model.go

```diff
@@ -0,0 +1,95 @@
+package feedist
+
+import (
+	"fmt"
+
+	"github.com/iov-one/weave"
+	"github.com/iov-one/weave/errors"
+	"github.com/iov-one/weave/orm"
+)
+
+var _ orm.CloneableData = (*Revenue)(nil)
+
+func (rev *Revenue) Validate() error {
+	if err := rev.Admin.Validate(); err != nil {
+		return errors.Wrap(err, "invalid admin signature")
+	}
+	if len(rev.Recipients) == 0 {
+		return errors.InvalidModelErr.New("no recipients")
+	}
+	for i, r := range rev.Recipients {
+		if err := r.Address.Validate(); err != nil {
+			return errors.Wrap(err, fmt.Sprintf("recipient %d address", i))
+		}
+		if r.Weight < 0 {
+			return errors.InvalidModelErr.New(fmt.Sprintf("recipient %d has negative weight", i))
+		}
+	}
+	return nil
+}
+
+func (rev *Revenue) Copy() orm.CloneableData {
+	cpy := &Revenue{
+		Admin:      copyAddr(rev.Admin),
+		Recipients: make([]*Recipient, len(rev.Recipients)),
+	}
+	for i := range rev.Recipients {
+		cpy.Recipients[i] = &Recipient{
+			Address: copyAddr(rev.Recipients[i].Address),
+			Weight:  rev.Recipients[i].Weight,
+		}
+	}
+	return cpy
+}
+
+func copyAddr(a weave.Address) weave.Address {
+	cpy := make(weave.Address, len(a))
+	copy(cpy, a)
+	return cpy
+}
+
+type RevenueBucket struct {
+	orm.Bucket
+	idSeq orm.Sequence
+}
+
+// NewRevenueBucket returns a bucket for managing revenues state.
+func NewRevenueBucket() *RevenueBucket {
+	b := orm.NewBucket("revenue", orm.NewSimpleObj(nil, &Revenue{}))
+	return &RevenueBucket{
+		Bucket: b,
+		idSeq:  b.Sequence("id"),
+	}
+}
+
+// Create adds given revenue instance to the store and returns the ID of the
+// newly inserted entity.
+func (b *RevenueBucket) Create(db weave.KVStore, rev *Revenue) (orm.Object, error) {
+	key := b.idSeq.NextVal(db)
+	obj := orm.NewSimpleObj(key, rev)
+	return obj, b.Bucket.Save(db, obj)
+}
+
+// Save persists the state of a given revenue entity.
+func (b *RevenueBucket) Save(db weave.KVStore, obj orm.Object) error {
+	if _, ok := obj.Value().(*Revenue); !ok {
+		return errors.InvalidModelErr.New(fmt.Sprintf("invalid type: %T", obj.Value()))
+	}
+	return b.Bucket.Save(db, obj)
```

Comment:
If we were paranoid, we would also Validate it here again

Reply:
[`orm.Bucket.Save`](https://github.com/iov-one/weave/blob/master/orm/bucket.go#L147) is validating the model.